### ISSUE TITLE
Is it possible to get values of convolution layers?

### ISSUE THREAD

**Issue by emrekavur** [Kamnitsask/deepmedic#106]
Hi,
First, thank you sharing the DeepMedic source code. It is such a great system.

I have a question about it. After I finished training process, I would like to examine the values of convolution layers. I know it is saved somewhere in the *.save file at output folder. I unzipped the file and extracted uncompressed .save file. I opened it via a text editor. However, there are some unknown characters as well as some parts of compiled function in hundred thousand lines. Can I reach the numerical values of kernels of convolution layers? 

Best regards.

**Comment by Kamnitsask** [comment]
Hello,

With "values of convolution layers" do you mean kernels (filters) or activations (feature maps)?

For activations/feature maps, there is some functionality in place. Search for "Saving feature maps" in the main documentation, and the "Feature maps" section of a testConfig.cfg.

I am actually guessing you meant the filters, which is less straightforward. This functionality is not provided, so you ll have to understand certain classes and write some script yourself:

The saved model is not simply a numpy array with the filters, its far more than this. It is actually an instance of the class ./deepmedic/neuralnet/cnn3d.py -> Cnn3d()
It is created first when calling deepmedic/frontEndModules/createNewModel.py from the deepMedicRun (view -newModel). It is then loaded, altered during training and saved every epoch when calling deepmedic/routines/training.py/do_training( ) , routine that is startedd with a ./deepMedicRun -train.

If you want to get  the weights, you must have a good look at the Cnn3d class, its following variables and the related classes:
cnn3d.pathways
deepmedic/neuralnet/pathways.py
deepmedic/neuralnet/layers.py
Especially, in the layers.py, the class Block and its attributes self.params and the ones following. You ll have to look for the way they are created to see what type is each (Theano Shared) variable (w, b, g, batch norm variables, activation function's variables etc) and how to get their value (get_value function, see theano docs). You are probably mostly interested in _W I guess.

In short, you must write a script that does the following:
# Load the model instance (refer to 
cnn_instance = deepmedic.neuralnet.utils.load_object_from_gzip_file(/absolute/path/to/saved/model)
# Get the pathway, layer and parameter that you want.
np_arr_with_w_params = cnn_instance.pathways[ 0 ].getLayers()[ index_of_layer_of_interest]._W.get_value() 
# Do stuff

As a reference for making such script, you can check:
deepmedic/frontEndModels/createNewMode.py -> L169-172 where a model is loaded to transfer its parameters from another
deepmedic/neuralnet/transferParameters.py -> how to get pathways, laters, params.

Hope the above helps.
